hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/rollback/MarkerBasedRollbackStrategy.java

```diff
@@ -73,54 +80,110 @@ public MarkerBasedRollbackStrategy(HoodieTable<?, ?, ?, ?> table, HoodieEngineCo
+      String partitionPath = FSUtils.getRelativePartitionPath(new Path(basePath), fullFilePath.getParent());
+      switch (type) {
+        case MERGE:
+        case CREATE:
+          HoodieBaseFile baseFileToDelete = new HoodieBaseFile(fullFilePath.toString());
+          String fileId = baseFileToDelete.getFileId();
+          String baseInstantTime = baseFileToDelete.getCommitTime();
+          return Pair.of(Pair.of(partitionPath, fileId),
+              new HoodieRollbackRequest(partitionPath, fileId, baseInstantTime,
+                  Collections.singletonList(fullFilePath.toString()),
+                  Collections.emptyMap()));
+        case APPEND:
+          HoodieRollbackRequest rollbackRequestForAppend = getRollbackRequestForAppend(partitionFilePath);
+          return Pair.of(Pair.of(partitionPath, rollbackRequestForAppend.getFileId()),
+              rollbackRequestForAppend);
+        default:
+          throw new HoodieRollbackException("Unknown marker type, during rollback of " + instantToRollback);
+      }
+    };
+  }
+
+  private SerializableBiFunction<HoodieRollbackRequest, HoodieRollbackRequest, HoodieRollbackRequest> getRollbackReqCombineFunction() {
+    return (rollbackReq1, rollbackReq2) -> {
+      List<String> filesToBeDeleted = new LinkedList<>();
+      filesToBeDeleted.addAll(rollbackReq1.getFilesToBeDeleted());
+      filesToBeDeleted.addAll(rollbackReq2.getFilesToBeDeleted());
+      final Comparator<HoodieLogFile> logFileComparator = HoodieLogFile.getLogFileComparator();
+      HoodieLogFile latestLogFile = null;
+      long latestLogFileLen = -1;
+
+      for (Map.Entry<String, Long> pathLengthPair : rollbackReq1.getLogBlocksToBeDeleted().entrySet()) {
+        HoodieLogFile candidateLogFile = new HoodieLogFile(pathLengthPair.getKey());
+        if (latestLogFile == null || logFileComparator.compare(latestLogFile, candidateLogFile) >= 0) {
+          latestLogFile = candidateLogFile;
+          latestLogFileLen = pathLengthPair.getValue();
+        }
+      }
 
+      for (Map.Entry<String, Long> pathLengthPair : rollbackReq2.getLogBlocksToBeDeleted().entrySet()) {
+        HoodieLogFile candidateLogFile = new HoodieLogFile(pathLengthPair.getKey());
+        if (latestLogFile == null || logFileComparator.compare(latestLogFile, candidateLogFile) >= 0) {
+          latestLogFile = candidateLogFile;
+          latestLogFileLen = pathLengthPair.getValue();
+        }
+      }
+      return new HoodieRollbackRequest(rollbackReq1.getPartitionPath(), rollbackReq1.getFileId(),
+          rollbackReq1.getLatestBaseInstant(), filesToBeDeleted,
+          latestLogFile == null ? Collections.emptyMap() :
+              Collections.singletonMap(latestLogFile.getPath().toString(), latestLogFileLen));
+    };
+  }
+
+  protected HoodieRollbackRequest getRollbackRequestForAppend(String markerFilePath) {
+    Path filePath = new Path(basePath, markerFilePath);
```

Comment:
correct me if my understanding is wrong. thought this will get simplified a lot when we start generating maker file per log file. 
Just from the marker file path name, can't we deduce the log file that needs to be rolled back?  
prior to this patch, we don't create maker file w/ log file info, and hence we had to deduce the latest log file. but now, with this patch, wouldn't that get simplified. 
can you help me understand.

Reply:
Hi nsivabalan, I think we actually get it simplified because we no longer need file Listing of log files anymore. However, the code looks more complex. There are two reasons.
1. For change of the way how marker file generation, I choose to keep backward compatible rather than regenerating markers in UpgradeAndDownGrade. 
2. Assuming this case, we generate two log file in one commit due to rollover. We have two log files and two markers. According to current rollback mechanism, LogFiles was treated as a stack of logBlocks rather than an array of log blocks. I need to append command block to the end of latest log file instead of inserting several command blocks into this stack though such inserting seems also works. So there is a lot of code to do marker comparison.